### ISSUE TITLE
Changes in launch_template resource doesn't trigger a change in the autoscaling resource

### ISSUE THREAD

**Issue by aishahar** [hashicorp/terraform#30227]
Hi, we recently updated our terraform version to `1.0.11` from `1.0.10` and we realized that whenever there's a change in the launch_template (image_id), the autoscaling resource will not get updated and will not show in the terraform plan stage until the next run when the version of launch template is updated. However, in version `1.0.10`, this feature is still working for us. Also, when we tried with the latest version `1.1.2`, it gave us the same `terraform plan` output as `1.0.11`.

v1.0.11
Output from terraform plan. 
``` 
# module.abc[0].aws_launch_template.this will be updated in-place
  ~ resource "aws_launch_template" "this" {
        id                      = "lt-123"
      ~ image_id                = "ami-123abc" -> "ami-234bcd"
        name                    = "abc"
        tags                    = {
            "component" = "abc"
        }
        # (11 unchanged attributes hidden)
        # (3 unchanged blocks hidden)
 }
```
Only after `terraform apply` is run, then the next `terraform plan` will show the change in autoscaling resource as the version of the launch template has changed.

v1.0.10
Output from terraform plan
``` 
# module.abc[0].aws_autoscaling_group.this must be replaced
+/- resource "aws_autoscaling_group" "this" {
      ~ arn                       = "xxx" -> (known after apply)
      ~ availability_zones        = [
          - "ap-southeast-1a",
          - "ap-southeast-1b",
          - "ap-southeast-1c",
        ] -> (known after apply)
      ~ default_cooldown          = 300 -> (known after apply)
      - enabled_metrics           = [] -> null
      ~ id                        = "abc-holy-kit20211221024948463300000001" -> (known after apply)
      - load_balancers            = [] -> null
      - max_instance_lifetime     = 0 -> null
      ~ name                      = "abc-holy-kit20211221024948463300000001" -> (known after apply)
      ~ name_prefix               = "abc-holy-kit" -> (known after apply) # forces replacement
      ~ service_linked_role_arn   = "xxx" -> (known after apply)
      - suspended_processes       = [] -> null
        tags                      = [
            {
                "key"                 = "Name"
                "propagate_at_launch" = "true"
                "value"               = "abc"
            }
        ]
        # (15 unchanged attributes hidden)


      ~ mixed_instances_policy {
          ~ instances_distribution {
              ~ on_demand_allocation_strategy            = "prioritized" -> (known after apply)
              ~ spot_instance_pools                      = 2 -> (known after apply)
                # (3 unchanged attributes hidden)
            }

          ~ launch_template {
              ~ launch_template_specification {
                  ~ launch_template_name = "abc" -> (known after apply)
                    # (2 unchanged attributes hidden)
                }

                # (2 unchanged blocks hidden)
            }
        }
        # (2 unchanged blocks hidden)
    }

  # module.abc[0].aws_launch_template.this will be updated in-place
  ~ resource "aws_launch_template" "this" {
        id                      = "lt-0e64d73984bf5df43"
      ~ image_id                = "ami-123abc" -> "ami-234bcd"
        name                    = "abc"
        tags                    = {
            "component" = "abc"
        }
        # (11 unchanged attributes hidden)
        # (3 unchanged blocks hidden)
    }

  # module.abc[0].random_pet.this must be replaced
+/- resource "random_pet" "this" {
      ~ id        = "holy-kit" -> (known after apply)
      ~ keepers   = { # forces replacement
          ~ "lt_name" = "abc-184" -> "abc-185"
        }
        # (2 unchanged attributes hidden)
    }
```

Do anyone faced the same issue?

**Comment by jbardin** [comment]
Hi @aishahar,

Thanks for filing the issue. Could you also add the relevant portions of the configuration that contain these resources? It would help to see how the values for each attribute are being derived, and make a minimal reproducible example. 

There were very few changes between v1.0.10 and v1.0.11, are you certain that all providers are also exact the same version between each test?

Thanks!

**Comment by apparentlymart** [comment]
Focusing on the `# forces replacement` annotations in the second output, it seems like what caused the plan to replace `aws_autoscaling_group.this` was the change to `random_pet.this.keepers.lt_name`, and _not_ the in-place change to the image ID:

```
# module.abc[0].aws_autoscaling_group.this must be replaced
+/- resource "aws_autoscaling_group" "this" {
      # ...
      ~ name_prefix               = "abc-holy-kit" -> (known after apply) # forces replacement
      # ...
    }

  # module.abc[0].random_pet.this must be replaced
+/- resource "random_pet" "this" {
      ~ id        = "holy-kit" -> (known after apply)
      ~ keepers   = { # forces replacement
          ~ "lt_name" = "abc-184" -> "abc-185"
        }
        # (2 unchanged attributes hidden)
    }
```

Since we don't have the configuration to refer to yet I can't be sure, but I'm inferring from context that the `name_prefix` argument is populated something like this:

```
  name_prefix = "abc-${random_pet.this.id}"
```

Therefore replacing the "random pet" causes its `id` to be `(known after apply)`, which in turn forces replacement of the autoscaling group.

The autoscaling group configuration doesn't include any direct reference to the image ID -- only indirectly through the launch template -- and so there's nothing in this configuration to tell the AWS provider it needs to replace the autoscaling group when the image ID changes.

Given that, it seems like Terraform is working as intended here, but that this configuration isn't correct for the desired goal. I'm not familiar enough with autoscaling group patterns in particular to give specific advice, but perhaps _one_ way to get the desired effect would be to add another entry to the `keepers` in the `random_pet` that refers to `aws_launch_template.this.image_id`, which would therefore force generating a new random pet name each time the image ID changes, and then in turn force recreating the autoscaling group with the resulting new prefix.

**Comment by jw-maynard** [comment]
@apparentlymart @jbardin Maybe this should be raised in the provider repo instead. Not sure. I can add my experience here. I'm using TF v1.1.2 and the latest version of the AWS provider. I have EKS node groups that use launch_templates to define the type of notes they use (under the hood node groups use ASGs. I now have noticed that I need to do 2 applies to update a node group. The first updates the LT. Then they second apply notices the LT has changed and updates the Node Group/ASG accordingly.

Here's some config snippets.

```hcl
resource "aws_launch_template" "eks_minion_node_groups" {
  for_each               = var.worker_node_group_configurations
  name_prefix            = "${aws_eks_cluster.cluster.name}-${substr(each.key, 0, 8)}-"
  image_id               = each.value.use_bottlerocket ? local.ami_id.bottlerocket : local.ami_id.al2
  ebs_optimized          = true
  instance_type          = each.value.instance_type
  key_name               = var.keyname
...
}

esource "aws_eks_node_group" "eks_minion_node_group" {
  for_each        = aws_launch_template.eks_minion_node_groups
  cluster_name    = aws_eks_cluster.cluster.name
  node_group_name = "${aws_eks_cluster.cluster.name}-${each.key}"
  node_role_arn   = aws_iam_role.minion_role_node_groups.arn
  subnet_ids      = aws_eks_cluster.cluster.vpc_config[0].subnet_ids

  launch_template {
    id      = each.value.id
    version = each.value.latest_version
  }
...
}
```

What I think might be a key to this issue, at least in my case, is that the `latest_version` part of the LT resource no longer shows as "known after apply" during a plan output. You can see that above in @aishahar output also.

Interestingly I found an issue in the provider repo for perhaps the "opposite" bug https://github.com/hashicorp/terraform-provider-aws/issues/19411

**Comment by apparentlymart** [comment]
Thanks for sharing that additional context, @jw-maynard!

That does indeed sound like a provider issue, though the other provider issue you linked suggests that it might be a pretty complicated scenario for the provider to model. Terraform Core's expectation here is that the provider would mark `latest_version` as `(known after apply)` in any situation where the provider expects that value to change during the apply step, and in hashicorp/terraform-provider-aws#19411 it seems like the provider previously did have logic for that but it was "too sensitive" in the sense that it was signalling a potential change even though there was no other planned change that would potentially cause that version update.

I'm not familiar with the mechanics of the AWS autoscaling API here so I don't know what exactly the right answer is, but I would agree that in your case it does seem like a provider bug if you were able to change `aws_launch_template` in a way that implies a new version without the provider marking `latest_version` as `(known after apply)` (or alternatively, as a new _known_ value if it's predictable somehow). Of course there's also the question of what `aws_eks_node_group` should do in response to seeing `version` change, but I gather from what you've described that you've observed that resource type correctly handling version changes before and so the problem does seem to be upstream in `aws_launch_template`.

Terraform Core includes a safety check that aims to catch some situations like this and report them, but these older resource types in the AWS provider are unfortunately exempt from that being a fatal error as a side-effect of some unrelated allowances we made to avoid making too much noise about some bugs in older versions of the Terraform SDK. However, if you run an operation like this with `TF_LOG=trace` enabled then you may see a note in the internal logs which warns about the detected inconsistency, and would thus confirm that this is a provider bug:

```
[WARN] Provider provider["registry.terraform.io/hashicorp/aws"] produced an unexpected new value for aws_launch_template.eks_minion_node_groups, but we are tolerating it because it is using the legacy plugin SDK.
      - .latest_version: was cty.NumberIntVal(1), but now cty.NumberIntVal(2)
```

(The absence of this warning doesn't _eliminate_ the possibility of it being a provider bug, but it would be a different sort of bug than the one I'm imagining. Other messages containing the string "we are tolerating it" may surface similar sorts of problems, but they also often also just describe quirks in the old Terraform SDK -- the reason why this is just a warning and not an error for this resource type -- so the mere _presence_ of them is not enough to indicate a bug. If you see ones that you aren't sure about and can share them here then we can analyze them a bit, though.)

**Comment by jw-maynard** [comment]
@apparentlymart Interestingly enough I added the `update_default_version` param to see if that would fix the issue and my apply failed with that exact error referencing the latest version. I'll report on the provider repo and link back to this discussion. Thanks!

**Comment by apparentlymart** [comment]
Aha! Thanks for the follow-up on that, @jw-maynard.

Looking at the other issue, I see that you saw a different variant of this reported downstream at the `aws_eks_node_group.eks_minion_node_group` resource, which was incorrectly blamed for producing an inconsistent final plan but really I expect there was an earlier trace log just like the on I showed, and indeed this is one of the "confusing errors from downstream operations" that the log message is referring to.

At this point this does feel like a provider-specific bug to me, rather than a Terraform Core bug, but we still have the pending request to see the configuration that motivated the original bug report here so I'm going to leave this open to give @aishahar a chance to respond to @jbardin's earlier question. It would also help to know about any messages from having the `TF_LOG=trace` environment variable set which include the string "we are tolerating it", which will help us to see whether the original problem has a similar root cause to what you just reported to the AWS provider team.

Thanks!

**Comment by aishahar** [comment]
Hey folks! Thanks for the prompt response.

During the terraform upgrade, we also perform terraform aws provider to 3.69.0 and realized that there's an open issue similar to the case I've listed above which is highlighted [here](https://github.com/hashicorp/terraform-provider-aws/issues/22015) and this issue will only be resolved in 3.71.0 which has yet to be released.
![image](https://user-images.githubusercontent.com/51870730/147177463-dd94b1aa-78c7-47b1-adb8-926b47882c9c.png)

Thus, I will be closing this issue as its not an issue pertaining to terraform but its due to terraform aws provider.

Thanks folks for helping out! 😄

**Comment by jw-maynard** [comment]
@aishahar Thanks so much for looping back with all this great information. I'll update the issue in the provider repo and when this is released I'll see if my issue is resolved. @apparentlymart Thanks for the in depth insight into the workings of Terraform. You're input in a thread is always insightful.
